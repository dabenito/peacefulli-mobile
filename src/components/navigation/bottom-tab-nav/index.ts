export { BottomTabNav } from "./bottom-tab-nav";
