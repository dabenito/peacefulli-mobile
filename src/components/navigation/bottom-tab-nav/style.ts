import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  text: {
    fontFamily: "pt_sans_bold",
    fontWeight: "bold",
  },
  tabIcon: {
    resizeMode: "center",
    height: 24,
    width: 24,
  },
  tabContainer: {
    paddingHorizontal: 30,
    paddingTop: 5,
  },
});
