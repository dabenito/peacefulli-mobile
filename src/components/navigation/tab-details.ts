import { ComponentClass } from "react";
import { Home } from "../home";
import { Support } from "../support";
import { Profile } from "../profile";
import { ListView } from "../list-view";
import { NavTabName } from "../../store/states/nav/nav-state";

export type TabDetails = {
  name: string;
  component: ComponentClass<any, any>;
  icon: string;
};

export class TabInfo {
  public static readonly SUPPORT: TabDetails = {
    name: NavTabName.SUPPORT,
    component: Support,
    icon: "hand-holding-heart",
  };
  public static readonly HOME: TabDetails = {
    name: NavTabName.HOME,
    component: Home,
    icon: "home",
  };
  public static readonly LIST: TabDetails = {
    name: NavTabName.LIST,
    component: ListView,
    icon: "list",
  };
  public static readonly PROFILE: TabDetails = {
    name: NavTabName.PROFILE,
    component: Profile,
    icon: "user-circle",
  };
}
