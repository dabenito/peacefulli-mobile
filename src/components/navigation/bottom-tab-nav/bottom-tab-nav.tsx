import React, { PureComponent } from "react";
import { connect } from "react-redux";
import AppLoading from "expo-app-loading";
import * as Font from "expo-font";
import Icon from "react-native-vector-icons/FontAwesome5";
import { createBottomTabNavigator } from "@react-navigation/bottom-tabs";
import { Color } from "../../../constants/color-constants";
import { styles } from "./style";
import { TabDetails, TabInfo } from "./tab-details";
import { customFonts } from "../../../constants/font-constants";
import { NestedNavAction } from "../../../store/actions/nav-action";
import { store } from "../../../store/redux-store";
import { RootState } from "../../../store/states/root-state";
import { NavTabName } from "../../../store/states/nav/nav-state";

const Tab = createBottomTabNavigator();

export interface BottomTabNavProps {
  defaultTab: string;
}

export interface BottomTabNavState {
  fontLoaded: boolean;
}

export class BottomTabNavUnconnected extends PureComponent<
  BottomTabNavProps,
  BottomTabNavState
> {
  constructor(props: BottomTabNavProps) {
    super(props);
    this.state = { fontLoaded: false };
  }

  create_tab(tabDetails: TabDetails): JSX.Element {
    return (
      <Tab.Screen
        name={tabDetails.name}
        component={tabDetails.component}
        listeners={{
          tabPress: () => {
            store.dispatch(NestedNavAction.tileChanged(tabDetails.name));
          },
        }}
        options={{
          tabBarLabel: tabDetails.name,
          tabBarIcon: ({ focused }) => {
            const color = focused ? Color.pink : Color.purple;
            return (
              <Icon
                name={tabDetails.icon}
                size={styles.tabIcon.height}
                color={color}
              />
            );
          },
        }}
      />
    );
  }

  render() {
    if (!this.state.fontLoaded) {
      return (
        <AppLoading
          startAsync={async () => {
            await Font.loadAsync(customFonts);
          }}
          onFinish={() => this.setState({ fontLoaded: true })}
          onError={console.warn}
        />
      );
    }
    return (
      <Tab.Navigator
        initialRouteName={this.props.defaultTab}
        tabBarOptions={{
          activeTintColor: Color.pink,
          inactiveTintColor: Color.purple,
          labelStyle: styles.text,
          style: styles.tabContainer,
        }}
      >
        {this.create_tab(TabInfo.LIST)}
        {this.create_tab(TabInfo.HOME)}
        {this.create_tab(TabInfo.SUPPORT)}
        {this.create_tab(TabInfo.PROFILE)}
      </Tab.Navigator>
    );
  }
}

const mapStateToProps = (state: RootState): BottomTabNavProps => {
  return {
    defaultTab: state.nestedNav.title ?? NavTabName.LIST,
  };
};

export const BottomTabNav = connect(mapStateToProps)(BottomTabNavUnconnected);
