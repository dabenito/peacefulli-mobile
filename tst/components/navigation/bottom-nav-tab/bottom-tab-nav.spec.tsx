import React from "react";
import { cleanup } from "@testing-library/react-native";
import { shallow, configure } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { BottomTabNavUnconnected } from "../../../../src/components/navigation/bottom-tab-nav/bottom-tab-nav";
import { TabInfo } from "../../../../src/components/navigation/bottom-tab-nav/tab-details";

console.error = () => {};
console.warn = () => {};

jest.mock("react-native-vector-icons/FontAwesome5", () => "Icon");

configure({ adapter: new Adapter() });

describe("BottomTabNav Tests", () => {
  afterEach(cleanup);

  it("loading nav with the LIST default", async () => {
    const nav = shallow(
      <BottomTabNavUnconnected defaultTab={TabInfo.LIST.name} />
    );
    nav.setState({ fontLoaded: true });
    expect(shallowToJson(nav)).toMatchSnapshot();
  });

  it("loading nav with the HOME default", async () => {
    const nav = shallow(
      <BottomTabNavUnconnected defaultTab={TabInfo.HOME.name} />
    );
    nav.setState({ fontLoaded: true });
    expect(shallowToJson(nav)).toMatchSnapshot();
  });

  it("loading nav with the SUPPORT default", async () => {
    const nav = shallow(
      <BottomTabNavUnconnected defaultTab={TabInfo.SUPPORT.name} />
    );
    nav.setState({ fontLoaded: true });
    expect(shallowToJson(nav)).toMatchSnapshot();
  });

  it("loading nav with the PROFILE default", async () => {
    const nav = shallow(
      <BottomTabNavUnconnected defaultTab={TabInfo.PROFILE.name} />
    );
    nav.setState({ fontLoaded: true });
    expect(shallowToJson(nav)).toMatchSnapshot();
  });

  it("font not loaded", async () => {
    const nav = shallow(
      <BottomTabNavUnconnected defaultTab={TabInfo.LIST.name} />
    );
    nav.setState({ fontLoaded: false });
    expect(shallowToJson(nav)).toMatchSnapshot();
  });
});
