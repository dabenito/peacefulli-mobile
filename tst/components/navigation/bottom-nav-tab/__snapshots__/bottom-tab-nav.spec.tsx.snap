// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomTabNav Tests font not loaded 1`] = `
<AppLoading
  onError={[Function]}
  onFinish={[Function]}
  startAsync={[Function]}
/>
`;

exports[`BottomTabNav Tests loading nav with the HOME default 1`] = `
<BottomTabNavigator
  initialRouteName="HOME"
  tabBarOptions={
    Object {
      "activeTintColor": "#FF00F0",
      "inactiveTintColor": "#6900FF",
      "labelStyle": Object {
        "fontFamily": "pt_sans_bold",
        "fontWeight": "bold",
      },
      "style": Object {
        "paddingHorizontal": 30,
        "paddingTop": 5,
      },
    }
  }
>
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="LIST"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "LIST",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="HOME"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "HOME",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="SUPPORT"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "SUPPORT",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="PROFILE"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "PROFILE",
      }
    }
  />
</BottomTabNavigator>
`;

exports[`BottomTabNav Tests loading nav with the LIST default 1`] = `
<BottomTabNavigator
  initialRouteName="LIST"
  tabBarOptions={
    Object {
      "activeTintColor": "#FF00F0",
      "inactiveTintColor": "#6900FF",
      "labelStyle": Object {
        "fontFamily": "pt_sans_bold",
        "fontWeight": "bold",
      },
      "style": Object {
        "paddingHorizontal": 30,
        "paddingTop": 5,
      },
    }
  }
>
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="LIST"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "LIST",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="HOME"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "HOME",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="SUPPORT"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "SUPPORT",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="PROFILE"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "PROFILE",
      }
    }
  />
</BottomTabNavigator>
`;

exports[`BottomTabNav Tests loading nav with the PROFILE default 1`] = `
<BottomTabNavigator
  initialRouteName="PROFILE"
  tabBarOptions={
    Object {
      "activeTintColor": "#FF00F0",
      "inactiveTintColor": "#6900FF",
      "labelStyle": Object {
        "fontFamily": "pt_sans_bold",
        "fontWeight": "bold",
      },
      "style": Object {
        "paddingHorizontal": 30,
        "paddingTop": 5,
      },
    }
  }
>
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="LIST"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "LIST",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="HOME"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "HOME",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="SUPPORT"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "SUPPORT",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="PROFILE"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "PROFILE",
      }
    }
  />
</BottomTabNavigator>
`;

exports[`BottomTabNav Tests loading nav with the SUPPORT default 1`] = `
<BottomTabNavigator
  initialRouteName="SUPPORT"
  tabBarOptions={
    Object {
      "activeTintColor": "#FF00F0",
      "inactiveTintColor": "#6900FF",
      "labelStyle": Object {
        "fontFamily": "pt_sans_bold",
        "fontWeight": "bold",
      },
      "style": Object {
        "paddingHorizontal": 30,
        "paddingTop": 5,
      },
    }
  }
>
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="LIST"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "LIST",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="HOME"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "HOME",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="SUPPORT"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "SUPPORT",
      }
    }
  />
  <Screen
    component={[Function]}
    listeners={
      Object {
        "tabPress": [Function],
      }
    }
    name="PROFILE"
    options={
      Object {
        "tabBarIcon": [Function],
        "tabBarLabel": "PROFILE",
      }
    }
  />
</BottomTabNavigator>
`;
